File: 面试题08.07.java
class Solution {
    public String[] permutation(String S) {
        StringBuffer s = new StringBuffer();
        List<String> ans = new ArrayList<String>();
        dfs(ans, S, s);
        return ans.toArray(new String[ans.size()]);
    }

    void dfs(List<String> l, String S, StringBuffer s) {
        if(S.length() == s.length()) {
            l.add(new String(s));
            return ;
        }
        for(int i = 0;i < S.length(); i ++) {
            String tmp = new String(s);
            if(tmp.contains(S.charAt(i) + "")) continue ;
            s.append(S.charAt(i));
            dfs(l, S, s);
            s.deleteCharAt(s.length() - 1);
        }
    }
}

File: 面试题08.07.go
func permutation(S string) []string {
	ans := make([]string, 0)
	var cur []byte
	visit := make([]bool, len(S))
	var dfs func()
	dfs = func() {
		if len(cur) == len(S) {
			ans = append(ans, string(cur))
		}
		for i := 0; i < len(S); i++ {
			if visit[i] == false {
				visit[i] = true
				cur = append(cur, S[i])
				dfs()
				visit[i] = false
				cur = cur[:len(cur)-1]
			}
		}
	}
	dfs()
	return ans
}